Dev (#1)
trailofbits/essence: README.md

# Essence

> This project aims to extract functions from llvm bitcode files and generate executables for them.

Reviewer: ```suggestion
This project aims to extract individual functions from LLVM bitcode files and generate executables for them.
```

Author: If we have a pure function that calls another pure function, then we currently extract both into the same new module. So I'm not sure if individual would be correct here.

Reviewer: Yeah, I see what you mean. Perhaps "aims to extract pure functions from LLVM bitcode [...]" and then a second sentence saying something like "multiple functions are extracted if a pure function contains calls to other pure functions."?

Author: I changed the wording to lean more on generating stand-alone executable instead of extracting functions since the extraction is actually just an implementation detail as we could have linked against the original bc module.

> Functions can be specified by name and convenient json input templates will be generated in which their arguments (including globals) can be specified.

Reviewer: ```suggestion
Functions can be specified by name and convenient JSON input templates will be generated in which their arguments (including globals) can be specified.
```

> The project is sub-devided into two components: `Essence` is the python tool which is meant for users and handles all of the end to end functionality, and `Handsanitizer` which is the c++ component which interacts directly with the LLVM bitcode modules and generates neccesary code/templates.

Reviewer: ```suggestion
The project is subdivided into two components: `Essence` is the Python tool which is meant for users and handles all of the end to end functionality, and `Handsanitizer` which is the C++ component which interacts directly with the LLVM bitcode modules and generates necessary code/templates.
```

(Also, is the capitalization on these tools here consistent with how they're used from the CLI?)

### About purity

> Our focus is primarly "pure" functions. There are two major categories of pure functions

Reviewer: ```suggestion
Our focus is primarly "pure" functions. There are two major categories of pure functions:
```

> 1. `read none`:

Reviewer: N.B.: You can use `1.` multiple times in Markdown to have the renderer perform auto-numbering for you. Not a big deal since we only have two elements here, but something to know!

   This category returns the same value for the same input every no matter how often the function is called.

>    Additionally a `read none` functions are not allowed to have side-effects and hence are not allowed to touch memory at all. 

Reviewer: Nit: they're allowed to touch *local* memory, as long as LLVM can prove that the local changes don't escape the function.

2. `write only`: These functions are similar to `read none` with the exception that these _are_ allowed to _write_ to memory. Since they are not allowed read from memory the behaviour should still be identical for every identical input, but this class can save for instance the return values to a global variable 

## Install instructions

```shell

> $ sudo apt install nlohmann-json3-dev

Reviewer: We should also vendor this dependency, since it's a single header file.

$ git clone --recursive https://github.com/trailofbits/essence
$ cd essence
$ cmake CMakeLists.txt
$ make 
$ pip install -e ./
```

### Commands  

> #### essence <input.bc> 

Reviewer: Use backticks for the headers here and below to prevent formatting confusion:

```suggestion
#### `essence <input.bc>`
```

This command lists the functions in a bitcode file, together with their signature and purity level 

#### essence --build [--output/-o outputdir] [--no-template] <input.bc> f1 f2 ...
This command will build an executable for f1 and f2.
The output directory can be specified, as well as that the json input template should not be (re)generated to preserve arguments inside the template file.

#### essence --build-read-none <input.bc>
Build all functions inside input.bc that are of purity level `read-none`

#### essence --build-write-only <input.bc>
Build all functions inside input.bc that are of purity level `write-only`

## Input routing / output routing
### Input 
To provide arguments to the function for which a standalone executable has been generated we provide convienent to use json input templates. These contain all global variables as well as a complete resprentation of the arguments possible arguments.

Essence also allows for setting the values underlying to values which point to memory.   

example:

> ```

Reviewer: ```suggestion
```c
```

typedef struct A {
   int v;
} A;

typedef struct B {
   int x;
   int* y;
   struct A* b;
   char * str;
} B;

void f(B b) { return; };

// json input template
{
  "globals": {
  },
  "arguments": {
    "barvictor": {
      "a0": int32_t,
      "a1": int32_t,
      "a2": {
        "a0": int32_t
      },
      "a3": [char]
    }
  }
}
```

If a value is a pointer we support the following inputs:
```json
{

>    "int_val_ptr_1": 30, # sets 30 at the addressed referenced by int_val_ptr,

Reviewer: N.B.: Try using `//` for comments here. JSON technically doesn't support comments at all, but `//` is accepted by more parsers than `#`.
